lib/Base/Organization/Collection.vala

```diff
@@ -16,192 +16,123 @@
-    }
-    return false;
+    // Update cache position
+    last_position_valid = true;
+    last_position = position;
+    last_iterator = iter;
+
+    // Return Object if one was found
+    return ! iter.is_end ()
+      ? iter.get () as Object
+      : null;
   }
 
   /**
-   * Compares two posts while sorting the list.
+   * Adds a new item into the list sorted.
    *
-   * The sorting is mostly reverse chronological, with a few exceptions.
-   * Any item that is not a Post, particular PseudoItems, will be placed on top.
-   * Also, if multiple posts that build a reply chain are found, the posts will
-   * be sorted chronological for a better understanding of the reply chain.
-   *
-   * @param a The first object to compare.
-   * @param b The second object to compare.
-   *
-   * @return How to sort both posts (positive when a before b, negative when b before a).
+   * @param item The new item to add.
    */
-  protected int compare_items (Object a, Object b) {
-    // Sort PseudoItems according to their index
-    if (a is PseudoItem && b is PseudoItem) {
-      // Retrieve the items
-      var pseudo_a = a as PseudoItem;
-      var pseudo_b = b as PseudoItem;
-
-      // Sort the items using the set index
-      uint x = pseudo_a.index;
-      uint y = pseudo_b.index;
-      return (int) (x > y) - (int) (x < y);
-    }
-
-    // Sort Post and not Post
-    if (a is Post && ! (b is Post)) {
-      return 1;
-    }
-    if (! (a is Post) && b is Post) {
-      return -1;
-    }
-
-    // Sort Posts
-    if (a is Post && b is Post) {
-      // Retrieve the posts
-      var post_a = a as Post;
-      var post_b = b as Post;
-
-      // Retrieve the direct parents for comparison
-      var a_parent = direct_parent (post_a);
-      var b_parent = direct_parent (post_b);
-
-      // Check if one post replies to the other
-      bool a_replied_b = a_parent.replied_to_id == b_parent.id;
-      bool b_replied_a = b_parent.replied_to_id == a_parent.id;
-      if (a_replied_b || b_replied_a) {
-        return (int) (a_replied_b) - (int) (b_replied_a);
-      }
+  protected void add_item (T item) {
+    SequenceIter iter = items.insert_sorted (item, sorter.compare_items);
+    uint position = iter.get_position ();
+    after_update (position, 0, 1);
+  }
 
-      // Otherwise sort by the date
-      DateTime x   = a_parent.creation_date;
-      DateTime y   = b_parent.creation_date;
-      int sort_mod = reverse_chronological ? -1 : 1;
-      return sort_mod * x.compare (y);
+  /*
+   * Run after an update to the list happened.
+   */
+  private void after_update (uint position, uint removed, uint added) {
+    // Invalidate cache if changes happened before cache position
+    if (position <= last_position) {
+      last_position_valid = false;
+      last_iterator = null;
+      last_position = 0;
     }
 
-    // If nothing fits, return 0
-    return 0;
+    // Send the update signal
+    items_changed (position, removed, added);
   }
```

Comment:
Should this be run after removing an item as well? There's only an `add_item()` method at the moment.

Reply:
Yes, this also needs to run when an item is removed. I just don't have a method to remove items yet, proof-of-concept and all.

`items_changed` is a signal from `ListModel` and notifies anyone using the list about changes to the list.